src/modules/sdexec/sdexec.c

```diff
@@ -0,0 +1,790 @@
+    /* If the unit reaches active.exited call StopUnit to cause stdout
+     * and stderr to reach eof, and the unit to transition to inactive.dead.
+     */
+    if (sdexec_unit_state (unit) == STATE_ACTIVE
+        && sdexec_unit_substate (unit) == SUBSTATE_EXITED
+        && proc->finished_response_sent) {
+
+        if (!proc->f_stop) {
+            flux_future_t *f;
+            flux_log (ctx->h, LOG_DEBUG, "stop %s", proc->name);
+            if (!(f = sdexec_stop_unit (ctx->h, ctx->rank, proc->name, "fail"))
+                || flux_future_then (f, -1, stop_continuation, proc) < 0) {
+                flux_log_error (ctx->h, "error initiating unit stop");
+                flux_future_destroy (f);
+                f = NULL;
+            }
+            proc->f_stop = f;
+        }
+    }
+    /* If the unit reaches failed.failed call ResetFailedUnit to cause stdout
+     * and stderr to reach eof, and the unit to transition to inactive.dead.
+     * We can land here for both a child failure and an exec failure.
+     * Start channel output here in case of the latter so it can be finalized.
+     */
+    if (sdexec_unit_state (unit) == STATE_FAILED
+        && sdexec_unit_substate (unit) == SUBSTATE_FAILED) {
+
+        sdexec_channel_start_output (proc->out);
+        sdexec_channel_start_output (proc->err);
+
+        if (!proc->f_stop) {
+            flux_future_t *f;
+            flux_log (ctx->h, LOG_DEBUG, "reset-failed %s", proc->name);
+            if (!(f = sdexec_reset_failed_unit (ctx->h, ctx->rank, proc->name))
+                || flux_future_then (f, -1, reset_continuation, proc) < 0) {
+                flux_log_error (ctx->h, "error initiating unit reset");
+                flux_future_destroy (f);
+                f = NULL;
+            }
+            proc->f_stop = f;
+        }
+    }
+    /* Conditionally send the final RPC response.
+     */
+    finalize_exec_request (ctx, proc, unit);
+}
+
+/* If a unit update matches a pending exec request, see if the request's
+ * state can be advanced.
+ */
+static void unit_cb (struct unit *unit, void *arg)
+{
+    struct sdexec_ctx *ctx = arg;
+    const flux_msg_t *msg;
+    struct sdproc *proc;
+
+    if ((msg = lookup_message_bypath (ctx->requests, sdexec_unit_path (unit)))
+        && (proc = flux_msg_aux_get (msg, "sdproc"))) {
+        update_pending_exec_request (ctx, proc, unit);
+    }
+}
+
+/* StartTransientUnit reply does not normally generate a sdexec.exec response,
+ * unless it fails.  Streaming responses continue  as "signals" are received.
+ */
+static void start_continuation (flux_future_t *f, void *arg)
+{
+    const flux_msg_t *msg = flux_future_aux_get (f, "request");
+    struct sdproc *proc = flux_msg_aux_get (msg, "sdproc");
+    struct sdexec_ctx *ctx = arg;
+
+    if (sdexec_start_transient_unit_get (f, NULL) < 0)
+        goto error;
+    sdexec_channel_close_fd (proc->in);
+    sdexec_channel_close_fd (proc->out);
+    sdexec_channel_close_fd (proc->err);
```

Comment:
could probably use small comment to explain why closing, natural thought would be that it stay open til the end.

Reply:
RIght, that's the systemd end of the fd which we know has been received (and duped) once a response to the StartTransientUnit request is received.  I'll add a comment.